net: multiple listen() events fail silently

eduardbcom opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

net
##### Description of change

<!-- Provide a description of the change below this comment. -->

Problem: It's possible to run listen() on a net.Server that's already listening to a port.
The result is silent failure, with the side effect of changing the _connectionKey and or _pipeName.

Solution: emit error if listen method called more than once.
close() method should be called between listen() method calls.

addaleax commented:
Related: #8294 … I read that as meaning that the current behaviour is the intended one?

eduardbcom commented:
This issue https://github.com/nodejs/node/issues/6190 for this PR still open.
Maybe documentation should be changed ?

Also, it should be more appropriate err code instead of `er.code = 'EADDRINUSE';`. Or we need to create one.

addaleax commented:
oh, yeah. /cc @trevnorris @jasnell

thefourtheye commented on lib/net.js:
```diff
@@ -1327,6 +1329,15 @@ function listen(self, address, port, addressType, backlog, fd, exclusive) {
+    var er = new Error('listen method has been called more than once.');
+    // maybe this is should be another code error
+    er.code = 'EADDRINUSE';
+    self.emit('error', er);
```
After emitting this error, why does it continue?

thefourtheye commented on test/parallel/test-net-server-listen-call-more-than-once.js:
```diff
@@ -0,0 +1,42 @@
+// This test calls listen function, and try to call it second time.
+// It should be error. listen function can be called just once.
+require('../common');
+var assert = require('assert');
```
Nit: You can use `const` here

thefourtheye commented on test/parallel/test-net-server-listen-call-more-than-once.js:
```diff
@@ -0,0 +1,42 @@
+server.on('error', function(e) {
+  console.error('server error');
+
+  // maybe this is should be another code error
```
Comment should be improved?

thefourtheye commented on test/parallel/test-net-server-listen-call-more-than-once.js:
```diff
@@ -0,0 +1,42 @@
+  console.error('server error');
+
+  // maybe this is should be another code error
+  if (e.code == 'EADDRINUSE') {
```
Use `===`

thefourtheye commented on test/parallel/test-net-server-listen-call-more-than-once.js:
```diff
@@ -0,0 +1,42 @@
+var assert = require('assert');
+var net = require('net');
+
+let server_listening_second_err = false;
```
Why two boolean flags? You can either use this or the next one, right?

thefourtheye commented on test/parallel/test-net-server-listen-call-more-than-once.js:
```diff
@@ -0,0 +1,42 @@
+var net = require('net');
+
+let server_listening_second_err = false;
+let server_listen_err = false;
```
Please use camelCasing to name variables.

eduardbcom commented:
@thefourtheye Thank you for comments.
Have changed PR according to your comments.

eduardbcom commented:
Hi there, @addaleax @thefourtheye @trevnorris @jasnell just a quick reminder.

jasnell commented:
Thanks for the reminder ping. I'll take another look at this a bit later on today.

eduardbcom commented:
Hi there, @addaleax @thefourtheye @trevnorris @jasnell just a quick reminder. Second one :)

thefourtheye commented:
@eduardbcom Can you please rebase, so that we can take another look at the latest code?

ronkorving commented:
It would be (have been?) nice if this could('ve) landed in Node 7.

jasnell commented:
As a semver-major it can't get into v7 without CTC review and ok. We can try. Adding it to the ctc agenda.

addaleax commented:
ping @eduardbcom … could you rebase this so that it can be ready for merging from your side? :)

eduardbcom commented:
@addaleax sorry for delay, I don't exactly understand what you expect to receive. rebase to master ? thx.

gibfahn commented:
@eduardbcom Yep, rebase against master to remove the merge conflicts github is complaining about.

rvagg commented on lib/net.js:
```diff
@@ -1327,6 +1329,15 @@ function listen(self, address, port, addressType, backlog, fd, exclusive) {
   var self = this;
 
+  if (self._listenHasBeenCalled) {
+    var er = new Error('listen method has been called more than once.');
```
are we using `let` in newly added code or are the existing `var`'s in here keeping us consistent?

addaleax commented:
From what I’ve gathered in the CTC meeting it seems like calling `.listen()` twice is currently problematic under basically all circumstances, so this LGTM and I’m +1 to landing on `v7.x`.

rvagg commented:
CTC meeting:

* @trevnorris raised that `_handle` is overwritten if you `listen()` a second time
* @addaleax raised a possible concern with #8294 which states in the docs that you can `listen()` on the same socket multiple times. It should probably say more clearly that you can only `listen()` on an already closed socket.
* There's also concern about whether `close()` is called in every situation when a socket is closed, in this PR it requires that path to unset the flag, is there any way a socket can be closed without that code path?

No negatives from the CTC meeting but it might need to come back next week or be handled async here.

jasnell commented:
I'm +1 on landing it in v7

jasnell commented:
@nodejs/ctc ... thoughts on this one?

Fishrock123 commented:
A quick search seems to indicate that `Server#close()` in the only place it can close on `Server` .

(Although that isn't true for `Socket` which closes it's handles in various places if an error occurs.)
(The places we care about all have `_handle.close()` I think.)

Fishrock123 commented:
I may have missed something though. :thinking: Someone may want to follow up and make sure server errors don't close it elsewhere.

rvagg commented:
That's the outstanding question for me, @indutny perhaps you're a good candidate for clarifying this for us? Is there any other way that a server can be closed that might bypass `Server#close()` such that this flag doesn't properly get reset. A socket close caused at the OS level perhaps?

Aside from that I don't think there's any other objection here.

trevnorris commented:
Let me clarify what currently happens on `master`. Source example:
```js
var s = net.createServer(function() {
  console.log(this.address());
}).listen(8001, 'localhost');

s.listen(8002);
```
Now run `nc localhost 8002` and you'll see that it binds to the later address. The reason for this is the asynchronous call to resolve the hostname in the first call, but the immediate call to bind to the port in the second call.

trevnorris commented:
Also note, this PR has a bug. If the `.listen()` is called, and the server errors (thus negating the need to call `.close()`), it should be possible to call `.listen()` again on the server object. This prevents that option.

jasnell commented:
Ping. What's the status on this one? As a semver-major, this needs to be wrapped up ASAP if there's going to be any chance of getting it into v7.0.0. We're a week away from the release and I do not want to land a semver-major at the last minute.

jasnell commented:
@nodejs/ctc ... I'm taking this off the v7.0.0 milestone. There are still outstanding issues, the PR needs to be rebased, it's semver-major, and I'm cutting the first release candidate today. This won't be able to make it in on time.

Trott commented:
Because a decision has been made about whether this is going in to v7.0, I'm going to remove the `ctc-agenda` label.

jasnell commented:
@eduardbcom .. still interested in pursuing this?

eduardbcom commented:
@jasnell yep
let me walk though the comments and adjust fix

eduardbcom commented:
Uh

Unintentionally closed task. Git closed it after I cleaned up my fork..Keep in touch, I will create new PR and add you to participants.

Thx.

gibfahn commented:
>Unintentionally closed task. Git closed it after I cleaned up my fork..Keep in touch, I will create new PR and add you to participants.

@eduardbcom I think if you push more commits to the branch GitHub *might* reopen the PR, if not as long as you link to the previous PR in the new one (and maybe @ some people as you said) that'd be great.

eduardbcom commented:
Thanks for a good comment @trevnorris. I've added one more test to cover this scenario.
prependListener listener can deal with that. I think so. Please comment on this part.

As you may see I've introduced new ELISTENCALLEDTWISE constant. I'm not sure about it. Feel free to comment on this. If it's okay we should add this one into nodejs constants.

Let me know if we don't use es6 syntax for the purpose of consistency.

Thanks.

@jasnell @gibfahn

gibfahn commented:
CI: https://ci.nodejs.org/job/node-test-commit/8254/

eduardbcom commented:
@gibfahn have updated PR.
The problem was with port 80. I relied on the assumption that we will receive EACCESS error running server on port 80. That's true, but not always and not for all OS. So I've changed test a little bit.

gibfahn commented:
CI 2: https://ci.nodejs.org/job/node-test-commit/8264/

eduardbcom commented:
@gibfahn can you help me understand what the problem with https://ci.nodejs.org/job/node-test-commit-arm/8099/ ? Thx.

gibfahn commented:
>@gibfahn can you help me understand what the problem with https://ci.nodejs.org/job/node-test-commit-arm/8099/ ? Thx.

There is a problem with the job status reporting for the arm jobs. If it's green when you go to ci.nodejs.org, then it's fine.

gibfahn commented:
@eduardbcom Could you add `Refs:` and `Fixes:` lines to the commit message? Edited message would look something like:

```
net: multiple listen() events fail silently

Problem:
It's possible to run listen() on a net.Server that's already listening
to a port. The result is silent failure, with the side effect of
changing the _connectionKey and or _pipeName.

Refs: https://github.com/nodejs/node/pull/8294
Fixes: https://github.com/nodejs/node/issues/6190
```

gibfahn commented:
Okay, this is rebased, passing CI, and has an LGTM from @addaleax. Let's see if we can get this landed. What's still outstanding from the below issues?

cc/ @rvagg @jasnell @indutny @addaleax @trevnorris @thefourtheye 

---

>>Also note, this PR has a bug. If the .listen() is called, and the server errors (thus negating the need to call .close()), it should be possible to call .listen() again on the server object. This prevents that option.
>
>Thanks for a good comment @trevnorris. I've added one more test to cover this scenario.
prependListener listener can deal with that. I think so. Please comment on this part.

>@trevnorris raised that _handle is overwritten if you listen() a second time

@trevnorris does this LGTY now?

---

>>>There's also concern about whether close() is called in every situation when a socket is closed, in this PR it requires that path to unset the flag, is there any way a socket can be closed without that code path?
>>
>>A quick search seems to indicate that `Server#close()` in the only place it can close on `Server` .
>>
>>(Although that isn't true for `Socket` which closes it's handles in various places if an error occurs.)
(The places we care about all have `_handle.close()` I think.)
>>
>>I may have missed something though. 🤔 Someone may want to follow up and make sure server errors don't close it elsewhere.
>
>That's the outstanding question for me, @indutny perhaps you're a good candidate for clarifying this for us? Is there any other way that a server can be closed that might bypass Server#close() such that this flag doesn't properly get reset. A socket close caused at the OS level perhaps?

Was this ever resolved (@indutny @rvagg)?

---

>@addaleax raised a possible concern with #8294 which states in the docs that you can listen() on the same socket multiple times. It should probably say more clearly that you can only listen() on an already closed socket.

Raised a separate issue: https://github.com/nodejs/node/issues/11685

addaleax commented on lib/net.js:
```diff
@@ -1339,6 +1341,18 @@ Server.prototype.listen = function() {
 
+  if (this._listenHasBeenCalled) {
+    var er = new Error('listen method has been called more than once.');
+    er.code = 'ELISTENCALLEDTWISE';
```
Should this be `TWICE`?

eduardbcom replied:
Yes, of course. Will fix that

eduardbcom commented:
I have noticed that every time we use this._handle.close() we set _handle to null by running this._handle = null;
So I've changed code a little bit to rely on this behaviour. I think it's a slightly better solution, cuz we don't need update _listenHasBeenCalled to false in several places.

@gibfahn please run CI on this

joyeecheung commented on lib/net.js:
```diff
@@ -1137,11 +1137,27 @@ function Server(options, connectionListener) {
+      if (val === null) {
+        this._listenHasBeenCalled = false;
+      }
+      this.__handle = val;
```
To really hide it from users you can use a internal symbol:

```
const handleSymbol = Symbol('handle');
...
this[handleSymbol] = val;
...
return this[handleSymbol];
```

The same can be done to `_listenHasBeenCalled` too if you want to hide it. Old properties probably can't be hidden this way in case people are already using it..but new properties can ;)

joyeecheung replied:
To really hide it from users you can use a internal symbol:

```
const handleSymbol = Symbol('handle');
...
this[handleSymbol] = val;
...
return this[handleSymbol];
```

The same can be done to `_listenHasBeenCalled` too if you want to hide it. Old properties probably can't be hidden this way in case people are already using it..but new properties can ;)

gibfahn commented:
CI 3: https://ci.nodejs.org/job/node-test-commit/8279/

eduardbcom commented:
@gibfahn have updated pr, please run CI
@joyeecheung thanks for good idea

gibfahn commented:
CI 4: https://ci.nodejs.org/job/node-test-commit/8282/

eduardbcom commented:
@gibfahn @addaleax @jasnell 

rebased against master. maybe we need to run CI one more time and do something with this commit.

we can remove stalled label cuz it's not.

gibfahn commented:
CI 5: https://ci.nodejs.org/job/node-test-commit/8384/

@nodejs/ctc this needs some review.

cjihrig commented on lib/net.js:
```diff
@@ -1379,6 +1397,14 @@ Server.prototype.listen = function() {
 
+  if (this[serverListenHasBeenCalled]) {
+    var er = new Error('listen method has been called more than once.');
+    er.code = 'ELISTENCALLEDTWICE';
```
I don't think we should create new error codes.

evanlucas replied:
I agree

eduardbcom replied:
What error code should I use ?

cjihrig replied:
We don't necessarily need to attach a code. If we really want one, `EISCONN`, `EADDRNOTAVAIL`, or maybe `EALREADY` are options.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,62 @@
+let err1 = false;
+let err2 = false;
+
+function case1() {
```
You don't need to use functions. You can create separate block scopes.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,62 @@
+function case1() {
+
+  const dummyServer = net.Server();
+  const server1 = net.Server();
```
You don't need `server1` and `server2` as different names. They are in different scopes. Just use `server` for both.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,62 @@
+  const dummyServer = net.Server();
+  const server1 = net.Server();
+
+  // run some server
```
Can you capitalize and punctuate comments. This might fit on a single line too.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,62 @@
+
+  // run some server
+  // in order to simulate EADDRINUSE error
+  dummyServer.listen(0, 'localhost', () => {
```
Please add `common.mustCall()` around this callback.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,62 @@
+
+  server2.on('error', common.mustCall(function(e) {
+    assert.strictEqual(e.code, 'ELISTENCALLEDTWICE');
+    err2 = true;
```
`err2` isn't needed. This listener should only be called once, and is guaranteed by `common.mustCall()`.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,62 @@
+        server1.close();
+      });
+    } else if (e.code === 'ELISTENCALLEDTWICE') {
+      err1 = true;
```
Instead of having `err1`, couldn't this just be an assertion?

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,62 @@
+case1();
+case2();
+
+process.on('exit', function() {
```
I think this whole block could go away.

evanlucas commented on doc/api/http.md:
```diff
@@ -729,8 +729,8 @@ This function is asynchronous. `callback` will be added as a listener for the
-*Note*: The `server.listen()` method may be called multiple times. Each
-subsequent call will *re-open* the server using the provided options.
+*Note*:
+The `server.listen()` method may be called one more time iff there was an error during the first *listen* call or you explicitly called `server.close()`. Otherwise, `ELISTENCALLEDTWICE` error would be emitted.
```
Can you break this line at 80 columns please?

evanlucas commented on doc/api/http.md:
```diff
@@ -745,8 +745,8 @@ Start a UNIX socket server listening for connections on the given `path`.
-*Note*: The `server.listen()` method may be called multiple times. Each
-subsequent call will *re-open* the server using the provided options.
+*Note*:
+The `server.listen()` method may be called one more time iff there was an error during the first *listen* call or you explicitly called `server.close()`. Otherwise, `ELISTENCALLEDTWICE` error would be emitted.
```
Can you break this line at 80 columns please?

evanlucas commented on doc/api/http.md:
```diff
@@ -781,8 +781,8 @@ parameter is 511 (not 512).
-*Note*: The `server.listen()` method may be called multiple times. Each
-subsequent call will *re-open* the server using the provided options.
+*Note*:
+The `server.listen()` method may be called one more time iff there was an error during the first *listen* call or you explicitly called `server.close()`. Otherwise, `ELISTENCALLEDTWICE` error would be emitted.
```
Can you break this line at 80 columns please?

eduardbcom commented:
@evanlucas @cjihrig @gibfahn @Trott 

thanks for comments. updated pr

bnoordhuis commented on doc/api/http.md:
```diff
@@ -729,8 +729,10 @@ This function is asynchronous. `callback` will be added as a listener for the
+*Note*:
+The `server.listen()` method may be called one more time iff there was an error
+during the first *listen* call or you explicitly called `server.close()`.
+Otherwise, `EALREADY` error would be emitted.
```
"Otherwise, an `EALREADY` error is emitted." - likewise below.

Also, I'd probably replace "may be called one more time " with "can be called again."

bnoordhuis commented on doc/api/net.md:
```diff
@@ -186,8 +186,9 @@ Note:
-  subsequent call will *re-open* the server using the provided options.
+* The `server.listen()` method may be called one more time iff there was an error
+during the first *listen* call or you explicitly called `server.close()`.
+Otherwise, `EALREADY` error would be emitted.
```
Can you indent here?

bnoordhuis commented on lib/net.js:
```diff
@@ -1174,11 +1176,27 @@ function Server(options, connectionListener) {
+      return this[serverHandle];
+    },
+    enumerable: true
+  });
```
I don't know how I feel about turning an internal property into an accessor...  Why did you decide on this approach?  Is it because of https://github.com/nodejs/node/pull/8419#issuecomment-251793465?

eduardbcom replied:
The main reason I decided to do it in this way it's because we need to set `serverListenHasBeenCalled` back to false when we close connection or some error has happened. And I have noticed that for both cases we call the same code to handle `_handle` field. Concretely:

`    this._handle.close();
    this._handle = null;` 

So in order to avoid repetitions and do something like this every time we close this._handle

 `   
    this._handle.close();
    this._handle = null;
    this.serverListenHasBeenCalled = false
` 

I decided to make this stuff in one place.

A big plus of this, is that if someone in a future will add some logic to handle error, they will not forget to set `this.serverListenHasBeenCalled = false`.

cjihrig replied:
@bnoordhuis what are you thoughts on this now? Still conflicted?

bnoordhuis replied:
I'd stick with simple properties.  Accessors have overhead.  If that means duplicating an assignment in two places, that's something I can live with.

bnoordhuis commented on lib/net.js:
```diff
@@ -1379,6 +1397,14 @@ Server.prototype.listen = function() {
 
+  if (this[serverListenHasBeenCalled]) {
+    var er = new Error('listen method has been called more than once.');
+    er.code = 'EALREADY';
```
This is in line with what libuv returns for client sockets but using EALREADY instead of EINVAL (the standard POSIX error for trying to rebind) is one of my little regrets there.

Thoughts, anyone?  Consistency or conformity?

cjihrig replied:
If you regret it, then we should probably go with `EINVAL`.

jasnell replied:
Agreed.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,46 @@
+  }));
+
+  server.listen(0, 'localhost', () => server.close());
+  server.listen(1, 'localhost');
```
Why `1` here?

eduardbcom replied:
Right, no reason for this.

Trott commented:
Since this has received three CTC reviews since application of `ctc-review` label, I'm going to go ahead and remove that label. If anyone thinks it should stay, by all means, re-apply it.

jasnell commented:
Ping @eduardbcom... some new comments have been added

eduardbcom commented:
@jasnell thanks for the reminder ping.

@bnoordhuis please check the comment for your question about the `_handle` property.

@cjihrig @evanlucas

updated pr. Thank you guys for your comments.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,48 @@
+    assert.strictEqual(e.code, 'EINVAL');
+  }));
+
+  server.listen(common.mustCall(
```
This can all be one line.

cjihrig commented on test/parallel/test-net-server-call-listen-multiple-times.js:
```diff
@@ -0,0 +1,48 @@
+// after first time is fail (without calling close method).
+
+// The second one is about calling a listen method twice.
+// It should be EALREADY but the first server listen call is working.
```
Should `EALREADY` be `EINVAL`?

cjihrig commented on lib/net.js:
```diff
@@ -1373,6 +1391,14 @@ Server.prototype.listen = function() {
   var options = normalized[0];
   const cb = normalized[1];
 
+  if (this[serverListenHasBeenCalled]) {
```
Can you check `=== true` here.

cjihrig commented on lib/net.js:
```diff
@@ -1168,11 +1170,27 @@ function Server(options, connectionListener) {
+  });
+
   this._handle = null;
+
```
Can you remove the unnecessary whitespace changes (here and in other places).

eduardbcom commented:
thanks @cjihrig, updated

eduardbcom commented:
@jasnell @bnoordhuis @Trott  @cjihrig @evanlucas quick reminder

bnoordhuis commented on doc/api/http.md:
```diff
@@ -729,8 +729,10 @@ This function is asynchronous. `callback` will be added as a listener for the
-*Note*: The `server.listen()` method may be called multiple times. Each
-subsequent call will *re-open* the server using the provided options.
+*Note*:
+The `server.listen()` method can be called again iff there was an error
```
I forgot what our position was on using 'iff' in the documentation: yay or nay?  `git grep` suggests nay.

jasnell replied:
My preference is to avoid it. Please use `if and only if` instead.
